File: oolang-parser/src/test/java/oo/parser/OolangParserTest.java
package oo.parser;

import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class OolangParserTest {
    @Test
    public void parseSimplestClass() {
        OolangParser parser = parserForCode("class A");
        OolangParser.ClassDeclarationContext classDecl = parser.classDeclaration();
        assertThat(classDecl.getChildCount()).isEqualTo(2);
        assertThat(classDecl.CLASS().getText()).isEqualTo("class");
        assertThat(classDecl.simpleIdentifier().getText())
                .isEqualTo("A");
        assertThat(classDecl.classBody()).isNull();
    }

    @Test
    public void parseSimplestClassWithBraces() {
        OolangParser parser = parserForCode("class A {}");
        OolangParser.ClassDeclarationContext classDecl = parser.classDeclaration();
        assertThat(classDecl.getChildCount()).isEqualTo(3);
        assertThat(classDecl.CLASS().getText()).isEqualTo("class");
        assertThat(classDecl.simpleIdentifier().getText())
                .isEqualTo("A");
        assertThat(classDecl.classBody().getText()).isEqualTo("{}");
    }

    @Test
    public void parseSimplestInterface() {
        OolangParser parser = parserForCode("interface B");
        OolangParser.ClassDeclarationContext classDecl = parser.classDeclaration();
        assertThat(classDecl.getChildCount()).isEqualTo(2);
        assertThat(classDecl.INTERFACE().getText()).isEqualTo("interface");
        assertThat(classDecl.simpleIdentifier().getText())
                .isEqualTo("B");
        assertThat(classDecl.classBody()).isNull();
    }

    private static OolangParser parserForCode(String code) {
        return new OolangParser(new CommonTokenStream(new OolangLexer(CharStreams.fromString(code))));
    }
}

File: oolang-parser/src/test/kotlin/oo/parser/OolangParserTest.kt
package oo.parser

import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream
import org.antlr.v4.runtime.Token
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class OolangParserTest {
    @Test
    fun parseSimplestClass() {
        val parser = parserForCode("class A")
        val classDecl = parser.classDeclaration()
        assertThat(classDecl.childCount).isEqualTo(2)
        assertThat(classDecl.CLASS().text).isEqualTo("class")
        assertThat(classDecl.simpleIdentifier().text)
            .isEqualTo("A")
        assertThat(classDecl.classBody()).isNull()
    }

    @Test
    fun parseSimplestClassWithBraces() {
        val parser = parserForCode("class A {}")
        val classDecl = parser.classDeclaration()
        assertThat(classDecl.childCount).isEqualTo(3)
        assertThat(classDecl.CLASS().text).isEqualTo("class")
        assertThat(classDecl.simpleIdentifier().text)
            .isEqualTo("A")
        assertThat(classDecl.classBody().text).isEqualTo("{}")
    }

    @Test
    fun parseSimplestInterface() {
        val parser = parserForCode("interface B")
        val classDecl = parser.classDeclaration()
        assertThat(classDecl.childCount).isEqualTo(2)
        assertThat(classDecl.INTERFACE().text).isEqualTo("interface")
        assertThat(classDecl.simpleIdentifier().text)
            .isEqualTo("B")
        assertThat(classDecl.classBody()).isNull()
    }

    private fun parserForCode(code: String) =
        OolangParser(CommonTokenStream(OolangLexer(CharStreams.fromString(code))))
}